### PR TITLE
Fixed the Tuning copying off of strings

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
The code in #2422 was buggy at my last commit, I rushed in some code to copy off candidate strings. That turned out to be broken. I've fixed it here

### CODE REVIEW

File: core/src/impl/Kokkos_Profiling.cpp [hunk 1]
```diff
@@ -112,10 +112,10 @@ void beginParallelFor(const std::string& kernelPrefix, const uint32_t devID,
     Experimental::begin_context(context_id);
     Experimental::VariableValue contextValues[] = {
         Experimental::make_variable_value(
-            Experimental::kernel_name_context_variable_id,
-            kernelPrefix.c_str()),
+            Experimental::kernel_name_context_variable_id, kernelPrefix),
         Experimental::make_variable_value(
-            Experimental::kernel_type_context_variable_id, "parallel_for")};
+            Experimental::kernel_type_context_variable_id,
+            std::string("parallel_for"))};
```
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.cpp:118]: There is no need for this change

File: core/src/impl/Kokkos_Profiling.cpp [hunk 2]
```diff
@@ -807,21 +809,20 @@ VariableValue make_variable_value(size_t id, double val) {
...
   return variable_value;
 }
-Experimental::VariableValue make_variable_value(size_t id, const char* val) {
+Experimental::VariableValue make_variable_value(size_t id,
+                                                const std::string& val) {
   VariableValue variable_value;
-  variable_value.type_id            = id;
-  variable_value.value.string_value = val;
+  variable_value.type_id = id;
+  strncpy(variable_value.value.string_value, val.c_str(), 63);
```
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.cpp:816]: Do not use magic numbers

File: core/src/impl/Kokkos_Profiling.cpp [hunk 3]
```diff
@@ -807,21 +809,20 @@ VariableValue make_variable_value(size_t id, double val) {
...
   return variable_value;
 }
-SetOrRange make_candidate_set(size_t size, const char** data) {
+SetOrRange make_candidate_set(size_t size, std::string* data) {
   SetOrRange value_set;
-  const char** data_copy = new const char*[size];
+  value_set.set.values.string_value = new tuning_string[size];
```
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.cpp:821]: I held back from commenting above for `variable_value.value.string_value` but `value_set.set.values.string_value` are you for real?

Comment (dalg24) [reply core/src/impl/Kokkos_Profiling.cpp:821]: Also would you remind me who is in charge for calling delete?

File: core/src/impl/Kokkos_Profiling.cpp (hunk 1 above)
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.cpp:118]: Building the tutorial exercise I'm working on, if I remove that I get linker errors:

```
/home/dzpolia/src/spack/opt/spack/linux-rhel7-skylake_avx512/gcc-7.1.0/binutils-2.34-tyakqk6dwdx2nd5ybnmfpayrszna6yj5/bin/ld: /home/dzpolia/src/kokkos-tutorials/Intro-Full/Exercises/tuning/Begin/libkokkos.a(Kokkos_Profiling.o): in function `Kokkos::Tools::beginParallelFor(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned int, unsigned long*)':
/ascldap/users/dzpolia/Kokkos/kokkos/core/src/impl/Kokkos_Profiling.cpp:113: undefined reference to `Kokkos::Tools::Experimental::make_variable_value(unsigned long, char const*)'
```

File: core/src/impl/Kokkos_Profiling.cpp (hunk 3 above)
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.cpp:821]: These guys live through the whole application. If we really wanted to, we could clean this up in finalize, maybe? But these generally do need to live at least through the end of Kokkos

File: core/src/impl/Kokkos_Profiling.cpp (hunk 2 above)
Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.cpp]: Yup, fixed that one via bd5eb14, good comment. There are magic numbers elsewhere in Profiling, I'll fix those in a separate PR though

File: core/src/impl/Kokkos_Profiling.cpp (hunk 1 above)
Comment (dalg24) [reply core/src/impl/Kokkos_Profiling.cpp:118]: https://github.com/kokkos/kokkos/blob/bd5eb149b6720c684df09bb6eacaa40b100c8051/core/src/impl/Kokkos_Profiling_Interface.hpp#L188

File: core/src/impl/Kokkos_Profiling.cpp [hunk 4]
```diff
@@ -807,21 +804,22 @@ VariableValue make_variable_value(size_t id, double val) {
...
 }
-SetOrRange make_candidate_set(size_t size, const char** data) {
+SetOrRange make_candidate_set(size_t size, std::string* data) {
   SetOrRange value_set;
-  const char** data_copy = new const char*[size];
+  value_set.set.values.string_value = new TuningString[size];
   for (size_t x = 0; x < size; ++x) {
-    data_copy[x] = new char[strnlen(data[x], 512)]{};
-    strncpy(const_cast<char*>(data_copy[x]), data[x], 512);
+    strncpy((value_set.set.values.string_value[x]), data[x].c_str(),
```
Comment (dalg24) [inline core/src/impl/Kokkos_Profiling.cpp:819]: why do you have that extra set of parentheses?

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_Profiling.cpp]: No good reason, I removed them in b6adc9e

**Comment by DavidPoliakoff** [comment]
Retest this please

**Comment by DavidPoliakoff** [comment]
Unless people have further comments this is ready to merge

**Comment by masterleinad** [comment]
@DavidPoliakoff I tried compiling locally with `TUNING=ON` and found some issues (and fixed them). Please have a look if you agree with the last commit.

**Comment by DavidPoliakoff** [comment]
@masterleinad  's changes look good to me, I think this can merge

File: core/src/impl/Kokkos_Profiling.cpp (hunk 3 above)
Comment (crtrott) [reply core/src/impl/Kokkos_Profiling.cpp]: Uhm please delete them at Kokkos::finalize. We have folks who check Kokkos for memory leaks.

**Pull request merged**